src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpStreamValidation.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Xml;
+using System.Xml.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using NuGet.Packaging;
+
+namespace NuGet.Protocol.Core.v3
+{
+    public static class HttpStreamValidation
+    {
+        public static void ValidateJObject(string uri, Stream stream, Action<JObject> validate)
+        {
+            try
+            {
+                using (var reader = new StreamReader(stream, Encoding.UTF8, false, 4096, true))
+                using (var jsonReader = new JsonTextReader(reader) { CloseInput = false })
+                {
+                    var jObject = JObject.Load(jsonReader);
```

Comment:
Loading a JObject can be expensive, it would be nice if this could be done with just the reader.